Implement HttpContent.writeTo() async API

lorban opened the pull request:
Implementation of the `HttpContent.writeTo()` async API.

Fixes #8790

lorban committed:
first shot at HttpContent.writeTo

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/HttpContent.java:
```diff
@@ -61,15 +62,16 @@ public interface HttpContent
+     * <p>Write this HTTP content to a {@link Content.Sink}.</p>
      */
-    ByteBuffer getByteBuffer();
+    void writeTo(Content.Sink sink, Callback callback);
```
Ah I see why we can't have request/response here!

But can we have variants that also take a ByteBufferPool.Sized?
We can't pass in a pool via the HttpContentFactory, as pools are possibly per connector, which means per request.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java:
```diff
@@ -702,28 +702,7 @@ private void sendData(Request request, Response response, Callback callback, Htt
-            content.release();
-            callback.failed(x);
-        }
+        content.writeTo(response, callback);
```
maybe
```suggestion
        content.writeTo(response, 
            new ByteBufferPool.Sized(
                request.getComponents().getByteBufferPool(),
                request.getConnectionMetaData().getHttpConfiguration().getOutputBufferSize(),
                request.getConnectionMetaData().getHttpConfiguration().isUseOutputDirectByteBuffers()
            ),
            callback);
```

Better yet, add a `request.getComponents().getSizedByteBufferPool()` that could return a pool per connection rather than making the wrapper every time?

lorban committed:
change HttpContent.writeTo() to expose a byte buffer pool to implementations

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@gregw There are two things I take from you previous comment:

 - `org.eclipse.jetty.server.Components` should expose `getSizedByteBufferPool()` _instead of_ `getByteBufferPool()` as its implementations should always have a default size. `IOResources` too should be modified to take a sized buffer pool where it expects the pool + size + directness triplet. This refactoring should be done independently in another PR.

 - Making `HttpContent.writeTo()` take a buffer pool as one of its arguments looks terrible from an API perspective: there is no fundamental reason why writing would require a buffer pool, and most implementations don't: only `ResourceHttpContent` needs one.
In my thinking, the sink fundamentally is linked to the channel that does hold a sized pool so we should just expose that by making `writeTo()` take a subinterface of `Content.Sink`. This is what I did and it doesn't look too bad, and definitely better than `writeTo()` taking a sized pool which I tried too.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Components.java:
```diff
@@ -23,8 +23,11 @@
  */
 public interface Components
 {
+    // TODO remove?
     ByteBufferPool getByteBufferPool();
 
+    ByteBufferPool.Sized getSizedByteBufferPool();
+
```
I think just have the one method that returns a Sized pool (but with javadoc saying what the sizes are).

```suggestion
    ByteBufferPool.Sized getByteBufferPool();
```

gregw replied:
I've changed my mind... we do need new methods because we have different sizes for input and output:

```suggestion
    ByteBufferPool getByteBufferPool();

    ByteBufferPool.Sized getInputByteBufferPool();
    ByteBufferPool.Sized getOutputByteBufferPool();
```

gregw replied:
See #12046 which includes an implementation of this and also shows that using it can greatly simplify code

lorban replied:
Agreed that this should be taken on in a different PR.

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/HttpContent.java:
```diff
@@ -83,6 +86,14 @@ default long getBytesOccupied()
+    /**
+     * A {@link Content.Sink} that exposes a pre-sized buffer pool.
+     */
+    interface Sink extends Content.Sink
+    {
+        ByteBufferPool.Sized getSizedByteBufferPool();
+    }
```
I'm a big -1 on this interface.   Firstly associating a pool with a sink has nothing really to do with `HttpContent`. Any `Content.Sink` might ultimately have a buffer associated with it and then be wrapped with something like `BufferedContentSink`, which then has it's own buffers

let me ponder a better solution.....

gregw commented:
@lorban I have created #12046 that improves the use of Sized ByteBufferPools.

lorban commented:
@gregw I like https://github.com/jetty/jetty.project/pull/12046 but it's not solving where should `ResourceHttpContent.writeTo()` should find its buffer pool.

Let me try something...

lorban committed:
change the way to expose a byte buffer pool to misc implementations

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add writeTo(Sink, offset, length, cb)

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
remove unneeded change

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into experiment/jetty-12.1.x/HttpContent-writeTo

lorban commented:
Here is a summary of the status of the `HttpContent` overhaul:

in progress:

 - add sized `ByteBufferPool`s in `Components` (https://github.com/jetty/jetty.project/pull/12046)
 - add `HttpContent.writeTo()` (the core of this PR)
 - no ByteBufferPool in the signature of `HttpContent.writeTo()` as only `ResourceHttpContent` needs it (another proposal in this PR)

still needed:

 - remove `HttpContent.release()` (only needed by `CachingHttpContentFactory`, it's too complex and should be rewritten)
 - javadoc `HttpContent`

optional:

 - Replace `ByteBufferPool` with sized `ByteBufferPool` everywhere a pool + size + directness is used (e.g.: `IOResources`)

lorban committed:
fix tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented:
> @gregw I like #12046 but it's not solving where should `ResourceHttpContent.writeTo()` should find its buffer pool.

@lorban if you want to suck #12046 into this PR and close it, I'm OK with that.

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferPool.java:
```diff
@@ -158,6 +158,14 @@ public RetainableByteBuffer acquire()
+    /**
+     * A marker interface for classes that expose a pre-sized buffer pool.
+     */
+    interface Holder
+    {
+        ByteBufferPool.Sized getSizedByteBufferPool();
+    }
```
This is a little better... but I still don't like that wrapping can hide it.

We have adopted a wrapper heavy approach, so marker interfaces do not work well, unless we can unwrap all the layers looking for the marker.

Perhaps we need a Content.Sink.Wrapper class and then a static method that looks through the Sink and any wrappers looking for a Holder?

gregw replied:
The other option is to just make getByteBufferPool() a method on both Content.Sink and Content.Source

lorban replied:
This issue a way more complex than I anticipated, and since the best is the enemy of the good, I'm giving up and passing the sized pool as a parameter to `writeTo()`.

We could eventually revisit this later on if we find the time, but I don't want to block this PR on this matter.

lorban committed:
use well-defined OutputStream to Sink adapter

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
- pass buffer pool to HttpContent.writeTo()
 - give up on HttpContent.writeTo() overloading
 - standardize offset and length default values

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
leave sized pool problematic for another PR

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
update comment

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
make sure InputStreamContentSource always uses heap buffers

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
> @lorban if you want to suck #12046 into this PR and close it, I'm OK with that.

I prefer to focus solely on `HttpContent.writeTo()` here as I fear merging the two would turn this PR into an omnibus one. Those two concerns can be worked on separately and in parallel so I've reverted any related changes from this PR.

lorban commented:
@gregw as I respond to one of your comments, the issue of not passing `ByteBufferPool.Sized` to `HttpContent.writeTo()` is way more complex than I anticipated, and since the best is the enemy of the good, I'm giving up and passing the sized pool as a parameter.

I've removed all changes that are not directly related to `HttpContent.writeTo()` so I think this PR is now ready for its review  and to eventually be merged.

lorban committed:
fix ee9 resource range tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
stop relying on HttpConfiguration's useOutputDirectByteBuffers and outputBufferSize to size the static resources' buffers

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
stop relying on HttpConfiguration's useOutputDirectByteBuffers and outputBufferSize to size the static resources' buffers

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
Latest update: after a conversation with @sbordet, he managed to convince me that the `HttpConfiguration`'s `outputBufferSize` and `useOutputDirectByteBuffers` are unrelated to the buffer size/directness the `ResourceHttpContent` should use.

`ResourceHttpContent` is the only `HttpContent` implementation that requires a sized buffer pool to use in `writeTo()`, other `HttpContent` implementations use buffers with freely decided size/directness.

So I've modified `ResourceHttpContentFactory` so that it takes a sized buffer pool that can then be used by `ResourceHttpContent`, the buffer size and directness being configured by the instantiators of `ResourceHttpContentFactory`, i.e.: `ResourceServlet` or `ResourceService` which allow configuring those parameters via some servlet init parameters and setters respectively.

So the final signature if the new async API finally has this elegant form:
```java
void writeTo(Content.Sink sink, long offset, long length, Callback callback);
```

gregw commented:
I'm not entirely convinced that the directness is unrelated. But am happy to have an elegant API and use configuration to align sizes and directness.

lorban committed:
refactor buffer pool init and usage

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix bugs related to offset/length checks + add support for mmap'ed file > 2GiB

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix javadoc

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java:
```diff
@@ -99,20 +100,18 @@ else if (_baseResource.isAlias())
     }
 
-    private ByteBufferPool getByteBufferPool(Context context)
+    private ByteBufferPool findByteBufferPool()
```
Feels weird to have to find the BBP, rather than just passing it in at construction time.

I'd prefer to have a `BBP.Sized` as a parameter to the constructor of `ResourceHandler`, as it is a necessary component.

This would have the benefit of removing the size/direct fields, as they are encapsulated in `BBP.Sized`.

gregw replied:
The way the ServerConnector handles this is to have a full constructor taking a buffer pool, plus one without that passes null.   If null is passed, then then the server is looked at for its buffer pool.

We should have the same logic here.

lorban replied:
I'm not against adding a constructor that expects a `ByteBufferPool.Sized`, but that would clash with the two properties I added (`_byteBufferSize` and `_useDirectByteBuffers`) that specify the size of the buffers when the pool is taken from the server. There would be some confusion between the size/directness of the properties and the ones of the sized pool.

Or we could add a constructor that takes a non-sized pool, but then the API makes you control the pool both via a constructor argument _and_ a couple properties; that's not very elegant and begs the question _why not passing pool + size + directness as ctor args_ which leads to _why not requiring a sized pool_ which sends us back to square 1.

gregw replied:
I would remove those two properties.  Just have an constructor that that takes a Sized pool.   If that is null then you look for a buffer pool in the server and you can default the buffer size and directness (or look for the first connector and take them from that).

lorban replied:
Sounds alright, let me do that.

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/CachingHttpContentFactory.java:
```diff
@@ -596,9 +595,9 @@ public Resource getResource()
+        public void writeTo(Content.Sink sink, long offset, long length, Callback callback)
         {
-            return null;
+            callback.succeeded();
```
Should this write EOF and then succeed the callback?

lorban replied:
Yes, I think this should. Let me change that.

gregw commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/content/CachingHttpContentFactory.java:
```diff
@@ -403,6 +388,20 @@ public String getKey()
+        {
+            try
+            {
+                sink.write(false, BufferUtil.slice(_buffer.getByteBuffer(), (int)offset, (int)length), callback);
```
Why is this not last?
There are significant efficiencies to be had by sending content with last==true, rather than sending the content and then an empty buffer with last==true.

If you need a writeTo without last==true for multi-part ranges, then I think you need another method taking the boolean (or always pass the boolean).

lorban replied:
You just spotted a mistake: the last flag should be set to `true` here (and in the FileMapping impls too).

Multi-part ranges already deal with this by manipulating the last flag, see the comment I added in the [ee9 ResourceService](https://github.com/jetty/jetty.project/blob/219d442b3d4904f4c2f784434558482ba9e394ba/jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ResourceService.java#L868-L873).

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into experiment/jetty-12.1.x/HttpContent-writeTo

lorban committed:
fix last flag

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix last flag

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
Followup PRs:

 - https://github.com/jetty/jetty.project/pull/12046
 - https://github.com/jetty/jetty.project/pull/12060
 - https://github.com/jetty/jetty.project/pull/12069

Once all those have been merged, the `HttpContent` rework can be considered done.

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java:
```diff
@@ -294,6 +304,22 @@ public void setBaseResourceAsString(String base)
+    /**
+     * @param byteBufferSize The size of the byte buffers used to serve static resources.
+     */
+    public void setByteBufferSize(int byteBufferSize)
+    {
+        _byteBufferSize = byteBufferSize;
+    }
+
+    /**
+     * @param useDirectByteBuffers whether to use direct byte buffers to serve static resources.
+     */
+    public void setUseDirectByteBuffers(boolean useDirectByteBuffers)
+    {
+        _useDirectByteBuffers = useDirectByteBuffers;
+    }
```
Remove these and just use an optional constructor injected Sized pool.

lorban replied:
Done.

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -267,20 +267,26 @@ static Content.Source from(ByteBufferPool.Sized byteBufferPool, InputStream inpu
          */
         static Content.Source from(ByteBufferPool.Sized byteBufferPool, InputStream inputStream, long offset, long length)
         {
+            if (offset > 0 && length != 0)
+            {
+                try
+                {
+                    inputStream.skip(offset - 1);
+                    if (inputStream.read() == -1)
+                        throw new IllegalArgumentException("Offset out of range");
+                }
+                catch (IOException e)
+                {
+                    throw new RuntimeIOException(e);
+                }
+            }
```
Maybe move this logic and the length handling below into `InputStreamContentSource`, so it can be used elsewhere.

lorban replied:
Sure, done.

gregw reviewed (changes requested):
Very close now....

lorban committed:
accept sized pool as ctor argument

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
move offset/length logic to InputStreamContentSource

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
remove deprecated methods + deprecate constructions silently not using a buffer pool

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.1.x' into experiment/jetty-12.1.x/HttpContent-writeTo
